File: codes/ALGraph/Python/ALGraph.py
from enum import Enum
import SinglyLinkedListModule as slm
import ArrayBaseStackModule as asm
import CircularQueueModule as cqm


class Vertex(Enum):
    A = 0
    B = 1
    C = 2
    D = 3
    E = 4
    F = 5
    G = 6
    H = 7
    I = 8
    J = 9

class ALGraph:
    def __init__(self, nv):
        self.numV = nv
        self.numE = 0
        self.adjList = []
        self.visitInfo = []

        for i in range(nv):
            self.adjList.append(slm.SinglyLinkedList())
            self.adjList[len(self.adjList) - 1].SetSortRule(lambda d1, d2: d1.value-d2.value)

            self.visitInfo.append(0)

    def VisitVertex(self, visitV):
        if self.visitInfo[visitV.value] is 0:
            self.visitInfo[visitV.value] = 1
            print(chr(visitV.value + 65), end=' ')
            return True
        return False

    def AddEdge(self, fromV, toV):
        self.adjList[fromV.value].LInsert(toV)
        self.adjList[toV.value].LInsert(fromV)

        self.numE += 1
    
    def ShowGraphEdgeInfo(self):
        for i, e in enumerate(self.adjList):
            print("{}와 연결된 정점:".format(chr(i + 65)), end=' ')

            data = self.adjList[i].LFirst()
            if data[0] is True:
                print(chr(data[1].value + 65), end=' ')
                data = self.adjList[i].LNext()
                while data[0] is True:
                    print(chr(data[1].value + 65), end=' ')
                    data = self.adjList[i].LNext()
            print()
    
    def DFShowGraphVertex(self, startV):
        stack = asm.ArrayStack()
        visitV = startV

        self.VisitVertex(visitV)
        stack.SPush(visitV)

        data = self.adjList[visitV.value].LFirst()
        while data[0] is True:
            visitFlag = False

            if self.VisitVertex(data[1]) is True:
                stack.SPush(visitV)
                visitV = data[1]
                visitFlag = True
            else:
                data = self.adjList[visitV.value].LNext()
                while data[0] is True:
                    if self.VisitVertex(data[1]) is True:
                        stack.SPush(visitV)
                        visitV = data[1]
                        visitFlag = True
                        break
                    data = self.adjList[visitV.value].LNext()
            
            if visitFlag is False:
                if stack.SIsEmpty() is True:    # 처음 정점으로 돌아왔다면
                    break
                else:
                    visitV = stack.SPop()       # 돌아갈 정점
            
            data = self.adjList[visitV.value].LFirst()

        self.visitInfo = [0] * self.numV
    
    def BFShowGraphVertex(self, startV):
        queue = cqm.CircularQueue()
        visitV = startV

        self.VisitVertex(visitV)

        data = self.adjList[visitV.value].LFirst()
        while data[0] is True:
            if self.VisitVertex(data[1]) is True:
                queue.Enqueue(data[1])
            
            data = self.adjList[visitV.value].LNext()
            while data[0] is True:
                if self.VisitVertex(data[1]) is True:
                    queue.Enqueue(data[1])
                data = self.adjList[visitV.value].LNext()
            
            if queue.QIsEmpty() is True:    # 모든 정점을 다 방문했다면
                break
            else:
                visitV = queue.Dequeue()    # 연결된 다른 정점에서 다시 시작
            
            data = self.adjList[visitV.value].LFirst()

        self.visitInfo = [0] * self.numV

File: codes/ALGraph/C/ALGraph.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include "ALGraph.h"
#include "SinglyLinkedList.h"
#include "ArrayBaseStack.h"
#include "CircularQueue.h"

int WhoIsPrecede(int data1, int data2);

void GraphInit(ALGraph* pg, int nv)
{
	int i;

	pg->adjList = (List*)malloc(sizeof(List) * nv);
	pg->numV = nv;
	pg->numE = 0;

	for (i = 0; i < nv; i++)
	{
		ListInit(&(pg->adjList[i]));
		SetSortRule(&(pg->adjList[i]), WhoIsPrecede);
	}

	pg->visitInfo = (int*)malloc(sizeof(int) * pg->numV);
	memset(pg->visitInfo, 0, sizeof(int) * pg->numV);
}

void GraphDestroy(ALGraph* pg)
{
	if (pg->adjList != NULL)
		free(pg->adjList);

	if (pg->visitInfo != NULL)
		free(pg->visitInfo);
}

void AddEdge(ALGraph* pg, int fromV, int toV)
{
	LInsert(&(pg->adjList[fromV]), toV);
	LInsert(&(pg->adjList[toV]), fromV);
	pg->numE += 1;
}

void ShowGraphEdgeInfo(ALGraph* pg)
{
	int i;
	int vx;

	for (i = 0; i < pg->numV; i++)
	{
		printf("%c와 연결된 정점: ", i + 65);

		if (LFirst(&(pg->adjList[i]), &vx))
		{
			printf("%c ", vx + 65);

			while (LNext(&(pg->adjList[i]), &vx))
				printf("%c ", vx + 65);
		}
		printf("\n");
	}
}

int WhoIsPrecede(int data1, int data2)
{
	if (data1 < data2)
		return 0;
	else
		return 1;
}

int VisitVertex(ALGraph* pg, int visitV)
{
	if (pg->visitInfo[visitV] == 0)		// 방문한적 없던 정점이면
	{
		pg->visitInfo[visitV] = 1;
		printf("%c ", visitV + 65);
		return TRUE;					// 방문 성공
	}
	return FALSE;						// 방문 실패
}

void DFShowGraphVertex(ALGraph* pg, int startV)
{
	Stack stack;
	int visitV = startV;
	int nextV;

	StackInit(&stack);
	VisitVertex(pg, visitV);
	SPush(&stack, visitV);

	while (LFirst(&(pg->adjList[visitV]), &nextV) == TRUE)
	{
		int visitFlag = FALSE;

		if (VisitVertex(pg, nextV) == TRUE)
		{
			SPush(&stack, visitV);
			visitV = nextV;
			visitFlag = TRUE;
		}
		else
		{
			while (LNext(&(pg->adjList[visitV]), &nextV) == TRUE)
			{
				if (VisitVertex(pg, nextV) == TRUE)
				{
					SPush(&stack, visitV);
					visitV = nextV;
					visitFlag = TRUE;
					break;
				}
			}
		}

		if (visitFlag == FALSE)
		{
			if (SIsEmpty(&stack) == TRUE)		// 처음 정점으로 돌아왔다면
				break;
			else
				visitV = SPop(&stack);			// 돌아갈 정점
		}
	}

	memset(pg->visitInfo, 0, sizeof(int) * pg->numV);
}

void BFShowGraphVertex(ALGraph* pg, int startV)
{
	Queue queue;
	int visitV = startV;
	int nextV;

	QueueInit(&queue);

	VisitVertex(pg, visitV);

	while (LFirst(&(pg->adjList[visitV]), &nextV) == TRUE)
	{
		if (VisitVertex(pg, nextV) == TRUE)
			Enqueue(&queue, nextV);

		while (LNext(&(pg->adjList[visitV]), &nextV) == TRUE)
		{
			if (VisitVertex(pg, nextV) == TRUE)
				Enqueue(&queue, nextV);
		}

		if (QIsEmpty(&queue) == TRUE)	// 모든 정점을 다 방문했다면
			break;
		else
			visitV = Dequeue(&queue);	// 연결된 다른 정점에서 다시 시작
	}

	memset(pg->visitInfo, 0, sizeof(int) * pg->numV);
}